std-bits/table/src/main/java/org/enso/table/data/index/MultiValueIndex.java

```diff
@@ -1,48 +1,84 @@
 package org.enso.table.data.index;
 
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 import org.enso.table.aggregations.Aggregator;
-import org.enso.table.data.column.builder.object.StringBuilder;
 import org.enso.table.data.column.builder.object.*;
+import org.enso.table.data.column.builder.object.StringBuilder;
 import org.enso.table.data.column.storage.Storage;
 import org.enso.table.data.table.Column;
 import org.enso.table.data.table.Table;
 import org.enso.table.data.table.problems.AggregatedProblems;
 import org.enso.table.data.table.problems.FloatingPointGrouping;
 
-import java.util.*;
-import java.util.stream.Collectors;
-import java.util.stream.IntStream;
-
 public class MultiValueIndex {
   private final int keyColumnsLength;
-  private final Map<MultiValueKey, List<Integer>> locs;
+  private final Map<? extends MultiValueKeyBase, List<Integer>> locs;
   private final AggregatedProblems problems;
 
   public MultiValueIndex(Column[] keyColumns, int tableSize, Comparator<Object> objectComparator) {
     this(keyColumns, tableSize, null, objectComparator);
   }
 
-  public MultiValueIndex(Column[] keyColumns, int tableSize, int[] ordering, Comparator<Object> objectComparator) {
+  public MultiValueIndex(
+      Column[] keyColumns, int tableSize, int[] ordering, Comparator<Object> objectComparator) {
     this.keyColumnsLength = keyColumns.length;
-    this.locs = ordering == null ? new HashMap<>() : new TreeMap<>();
     this.problems = new AggregatedProblems();
+    this.locs =
+        ordering == null
+            ? buildUnorderedIndex(keyColumns, tableSize)
+            : buildOrderedIndex(keyColumns, tableSize, ordering, objectComparator);
+  }
 
+  private HashMap<UnorderedMultiValueKey, List<Integer>> buildUnorderedIndex(
```

Comment:
I agree the logic probably won't be more complicated but prefer avoiding repeating ourselves if possible.

Reply:
It won't give it back - because we still need to assign it to `locs` which has to be a common supertype - so we need to generalize the type to `Map<MultiValueKeyBase, List<Integer>>` - so we have no typesafety because then we can add either ordered or unordered keys and the static analysis cannot prove that we do this correctly.
But I guess it's ok - I'd just refactor `ordering == null` into a helper variable `boolean isOrdered` and then it should be quite clear in the code to see if we use it consistently.

I still personally prefer a little duplication in this case, but I appreciate your point that we want to limit it, so i think the `isOrdered` approach may be a decent middle ground.

Oh, I also wanted to avoid creating the `keyFactory` closure as it felt to me that it would be slower than a call to a regular method. But that is I think a premature micro-optimization, so I probably worried about it unnecessarily.